### PR TITLE
Documentation for Goal conditioning

### PR DESCRIPTION
**Pull request by vincentpierre**
### Proposed change(s)

Documentation for Goal conditioning

### Types of change(s)

- [ ] Bug fix
- [ ] New feature
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/main/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/main/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/main/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: docs/Training-Configuration-File.md [hunk 1]
```diff
@@ -43,6 +43,7 @@ choice of the trainer (which we review on subsequent sections).
...
+| `network_settings -> conditioning_type`       | (default = `hyper`) Conditioning type for the policy using goal observations. <br><br> `none` treats the goal observations as regular observations, `hyper` (default) uses a HyperNetowk with goal observations as input to generate some of the weights of the policy. Note that when using `hyper` the number of parameters of the network increases greatly. Therefore, it is recommended to reduce the number of `hidden_units` when using this `conditioning_type`
```
Comment (awjuliani) [inline docs/Training-Configuration-File.md:46]:
```suggestion
| `network_settings -> conditioning_type`       | (default = `hyper`) Conditioning type for the policy using goal observations. <br><br> `none` treats the goal observations as regular observations, `hyper` (default) uses a HyperNetwork with goal observations as input to generate some of the weights of the policy. Note that when using `hyper` the number of parameters of the network increases greatly. Therefore, it is recommended to reduce the number of `hidden_units` when using this `conditioning_type`
```

File: docs/Learning-Environment-Design-Agents.md [hunk 2]
```diff
@@ -560,6 +562,35 @@ between -1 and 1.
  of an entity to the `BufferSensor`.
  - Normalize the entities observations before feeding them into the `BufferSensor`.
 
+### Goal Observations
+
+It is possible for agents to collect observations that will be treated as "goal".
```
Comment (awjuliani) [inline docs/Learning-Environment-Design-Agents.md:567]:
```suggestion
It is possible for agents to collect observations that will be treated as "goal."
```

File: docs/Learning-Environment-Design-Agents.md [hunk 3]
```diff
@@ -560,6 +562,35 @@ between -1 and 1.
...
+A goal is used to condition the policy of the Agent, meaning that if the goal
+changes, the behavior of the Agent will change as well. Note that this is true
```
Comment (awjuliani) [inline docs/Learning-Environment-Design-Agents.md:569]: I think we should say that goals change the policy, which is a mapping from (observation -> action)

**Review by awjuliani — COMMENTED** [review]
Do we want to refer to it as a "goal signal" instead of a goal?

File: docs/Learning-Environment-Design-Agents.md [hunk 4]
```diff
@@ -560,6 +562,36 @@ between -1 and 1.
  of an entity to the `BufferSensor`.
  - Normalize the entities observations before feeding them into the `BufferSensor`.
 
+### Goal Signals
+
+It is possible for agents to collect observations that will be treated as "goal."
```
Comment (awjuliani) [inline docs/Learning-Environment-Design-Agents.md:567]:
```suggestion
It is possible for agents to collect observations that will be treated as a "goal."
```

Comment (awjuliani) [reply docs/Learning-Environment-Design-Agents.md:567]: I think within this paragraph we should also call these goal signals, and not goals.

**Review by awjuliani — APPROVED** [review]

File: docs/Learning-Environment-Design-Agents.md [hunk 5]
```diff
@@ -560,6 +562,36 @@ between -1 and 1.
  of an entity to the `BufferSensor`.
  - Normalize the entities observations before feeding them into the `BufferSensor`.
 
+### Goal Signals
+
+It is possible for agents to collect observations that will be treated as "goal signals".
+A goal signal is used to condition the policy of the Agent, meaning that if the goal
```
Comment (ervteng) [inline docs/Learning-Environment-Design-Agents.md:568]:
```suggestion
A goal signal is used to condition the policy of the agent, meaning that if the goal
```

File: docs/Learning-Environment-Design-Agents.md [hunk 6]
```diff
@@ -560,6 +562,36 @@ between -1 and 1.
...
+changes, the policy (i.e. the mapping from observations to actions) will change
+as well. Note that this is true
+for any observation since all observations influence the policy of the Agent to
+some degree. But by specifying a goal signal explicitly, we can make this conditioning
+more important to the agent. This feature can be used in settings where an agent
+must learn to solve different tasks that are similar by some aspects because the
+agent will learn to reuse learnings from different tasks to generalize better.
+In Unity, you can specify that a `VectorSensor` or
+a `CameraSensor` is a goal by attaching a `VectorSensorComponent` or a
+`CameraSensorComponent` to the Agent can selecting `Goal` as `Observation Type`.
```
Comment (vincentpierre) [inline docs/Learning-Environment-Design-Agents.md:578]:
```suggestion
`CameraSensorComponent` to the Agent and selecting `Goal Signal` as `Observation Type`.
```

File: docs/Learning-Environment-Design-Agents.md [hunk 7]
```diff
@@ -560,6 +562,36 @@ between -1 and 1.
  of an entity to the `BufferSensor`.
  - Normalize the entities observations before feeding them into the `BufferSensor`.
 
+### Goal Signals
```
Comment (vincentpierre) [inline docs/Learning-Environment-Design-Agents.md:565]:
```suggestion
### Goal Signal
```

File: docs/Learning-Environment-Design-Agents.md [hunk 8]
```diff
@@ -19,6 +19,8 @@
     - [RayCast Observation Summary & Best Practices](#raycast-observation-summary--best-practices)
   - [Variable Length Observations](#variable-length-observations)
     - [Variable Length Observation Summary & Best Practices](#variable-length-observation-summary--best-practices)
+  - [Goal Signals](#goal-signals)
```
Comment (vincentpierre) [inline docs/Learning-Environment-Design-Agents.md:22]:
```suggestion
  - [Goal Signal](#goal-signal)
```

File: docs/Learning-Environment-Design-Agents.md [hunk 9]
```diff
@@ -560,6 +562,36 @@ between -1 and 1.
  of an entity to the `BufferSensor`.
  - Normalize the entities observations before feeding them into the `BufferSensor`.
 
+### Goal Signals
+
+It is possible for agents to collect observations that will be treated as "goal signals".
```
Comment (vincentpierre) [inline docs/Learning-Environment-Design-Agents.md:567]:
```suggestion
It is possible for agents to collect observations that will be treated as "goal signal".
```

File: docs/Learning-Environment-Design-Agents.md [hunk 10]
```diff
@@ -560,6 +562,36 @@ between -1 and 1.
...
+`CameraSensorComponent` to the Agent can selecting `Goal` as `Observation Type`.
+On the trainer side, there are two different ways to condition the policy. This
+setting is determined by the
+[conditioning_type parameter](Training-Configuration-File.md#common-trainer-configurations).
+If set to `hyper` (default) a [HyperNetwork](https://arxiv.org/pdf/1609.09106.pdf)
+will be used to generate some of the
+weights of the policy using the goal observations as input. Note that using a
+HyperNetwork requires a lot of computations, it is recommended to use a smaller
+number of hidden units in the policy to alleviate this.
+If set to `none` the goal signals will be considered as regular observations.
```
Comment (vincentpierre) [inline docs/Learning-Environment-Design-Agents.md:587]:
```suggestion
If set to `none` the goal signal will be considered as regular observations.
```

**Review by ervteng — APPROVED** [review]

**Pull request merged**